File: src/slashDispatch.ts
import type { StateStore } from "src/state/StateStore";
import { shouldRedact } from "src/state/secretRedactor";
import { collectSecretKeys } from "src/tools/shared/secretKeys";
import type { StackDefinition, StackSummary } from "src/types/stack";
import { stringify as stringifyYaml } from "yaml";

export interface SlashDispatchContext {
  cwd: string;
  stateStore: StateStore;
}

export type DirectSlashResult = { ok: true; text: string } | { ok: false; error: string };

export function formatStacksTable(stacks: StackSummary[]): string {
  if (stacks.length === 0) {
    return "**Managed stacks**\n\nNo stacks defined under `.docker-agent/stacks/`.";
  }
  const header = "| Name | Services | Last applied |";
  const sep = "| --- | --- | --- |";
  const rows = stacks.map((s) => `| ${s.name} | ${s.serviceCount} | ${s.lastApplied ?? "never"} |`);
  return ["**Managed stacks**", "", header, sep, ...rows].join("\n");
}

function redactStackForDisplay(def: StackDefinition): StackDefinition {
  const clone = structuredClone(def);
  for (const spec of Object.values(clone.services)) {
    if (!spec.environment) continue;
    for (const [key, value] of Object.entries(spec.environment)) {
      if (shouldRedact(key)) spec.environment[key] = "***";
    }
  }
  return clone;
}

export function dispatchStacks(ctx: SlashDispatchContext): string {
  return formatStacksTable(ctx.stateStore.list());
}

export function dispatchYaml(stackName: string, ctx: SlashDispatchContext): DirectSlashResult {
  const def = ctx.stateStore.read(stackName);
  if (!def) {
    return { ok: false, error: `stack ${stackName} not found` };
  }
  const yaml = stringifyYaml(redactStackForDisplay(def));
  return { ok: true, text: `\`\`\`yaml\n${yaml.trimEnd()}\n\`\`\`` };
}

export function destroyStackPrompt(stackName: string, removeVolumes = false): string {
  return `Destroy stack ${stackName}${removeVolumes ? " with volumes" : ""}`;
}

export function isDestroyAllPrompt(content: string): boolean {
  return content.trim().toLowerCase() === "destroy all stacks";
}

export function parseDirectDestroyStack(
  content: string,
): { stackName: string; removeVolumes: boolean } | null {
  const trimmed = content.trim();
  const patterns = [
    /^Destroy stack (\S+)(?:\s+with volumes)?$/i,
    /^destroy (\S+)(?:\s+with volumes)?$/i,
  ];
  for (const pattern of patterns) {
    const match = trimmed.match(pattern);
    if (!match?.[1] || match[1].toLowerCase() === "all") continue;
    return {
      stackName: match[1],
      removeVolumes: /\swith volumes$/i.test(trimmed),
    };
  }
  return null;
}

export function dispatchSecretsList(
  stackName: string,
  ctx: SlashDispatchContext,
): DirectSlashResult {
  const def = ctx.stateStore.read(stackName);
  if (!def) {
    return { ok: false, error: `stack ${stackName} not found` };
  }
  const keys = [...collectSecretKeys(stackName, ctx)].sort();
  if (keys.length === 0) {
    return { ok: true, text: `No secret keys tracked for stack **${stackName}**.` };
  }
  const lines = keys.map((key) => `- ${key}`);
  return { ok: true, text: [`Secret keys for **${stackName}**:`, ...lines].join("\n") };
}

File: src/docker_agent/slash_dispatch.py
"""Slash command dispatch helpers.

Parity: ``src/slashDispatch.ts``.
"""

from __future__ import annotations

import copy
import re
from dataclasses import dataclass
from typing import Literal, TypedDict

import yaml

from docker_agent.state.secret_redactor import should_redact
from docker_agent.state.state_store import StateStore
from docker_agent.tools.shared.secret_keys import SecretKeysContext, collect_secret_keys
from docker_agent.types.stack import StackDefinition, StackSummary


@dataclass
class SlashDispatchContext:
    cwd: str
    state_store: StateStore


class DirectSlashOk(TypedDict):
    ok: Literal[True]
    text: str


class DirectSlashErr(TypedDict):
    ok: Literal[False]
    error: str


DirectSlashResult = DirectSlashOk | DirectSlashErr


def format_stacks_table(stacks: list[StackSummary]) -> str:
    if not stacks:
        return "**Managed stacks**\n\nNo stacks defined under `.docker-agent/states/`."
    header = "| Name | Services | Last applied |"
    sep = "| --- | --- | --- |"
    rows = [
        f"| {s.name} | {s.service_count} | {s.last_applied or 'never'} |"
        for s in stacks
    ]
    return "\n".join(["**Managed stacks**", "", header, sep, *rows])


def _redact_stack_for_display(definition: StackDefinition) -> StackDefinition:
    clone = copy.deepcopy(definition)
    for spec in clone.services.values():
        if not spec.environment:
            continue
        for key in list(spec.environment.keys()):
            if should_redact(key):
                spec.environment[key] = "***"
    return clone


def dispatch_stacks(ctx: SlashDispatchContext) -> str:
    return format_stacks_table(ctx.state_store.list())


def dispatch_yaml(stack_name: str, ctx: SlashDispatchContext) -> DirectSlashResult:
    definition = ctx.state_store.read(stack_name)
    if definition is None:
        return {"ok": False, "error": f"stack {stack_name} not found"}
    redacted = _redact_stack_for_display(definition)
    data = redacted.model_dump(by_alias=True, exclude_none=True)
    text = yaml.safe_dump(data, sort_keys=False)
    return {"ok": True, "text": f"```yaml\n{text.strip()}\n```"}


def destroy_stack_prompt(stack_name: str, remove_volumes: bool = False) -> str:
    suffix = " with volumes" if remove_volumes else ""
    return f"Destroy stack {stack_name}{suffix}"


def is_destroy_all_prompt(content: str) -> bool:
    return content.strip().lower() == "destroy all stacks"


def parse_direct_destroy_stack(content: str) -> dict[str, str | bool] | None:
    trimmed = content.strip()
    patterns = [
        re.compile(r"^Destroy stack (\S+)(?:\s+with volumes)?$", re.IGNORECASE),
        re.compile(r"^destroy (\S+)(?:\s+with volumes)?$", re.IGNORECASE),
    ]
    for pattern in patterns:
        match = pattern.match(trimmed)
        if not match or not match.group(1) or match.group(1).lower() == "all":
            continue
        return {
            "stack_name": match.group(1),
            "remove_volumes": bool(re.search(r"\swith volumes$", trimmed, re.IGNORECASE)),
        }
    return None


def dispatch_secrets_list(stack_name: str, ctx: SlashDispatchContext) -> DirectSlashResult:
    definition = ctx.state_store.read(stack_name)
    if definition is None:
        return {"ok": False, "error": f"stack {stack_name} not found"}
    keys_ctx = SecretKeysContext(cwd=ctx.cwd, state_store=ctx.state_store)
    keys = sorted(collect_secret_keys(stack_name, keys_ctx))
    if not keys:
        return {"ok": True, "text": f"No secret keys tracked for stack **{stack_name}**."}
    lines = [f"Secret keys for **{stack_name}**:"]
    lines.extend(f"- {key}" for key in keys)
    return {"ok": True, "text": "\n".join(lines)}


__all__ = [
    "DirectSlashResult",
    "SlashDispatchContext",
    "destroy_stack_prompt",
    "dispatch_secrets_list",
    "dispatch_stacks",
    "dispatch_yaml",
    "format_stacks_table",
    "is_destroy_all_prompt",
    "parse_direct_destroy_stack",
]